cobalt/base/process/process_metrics_helper_test.cc

```diff
@@ -0,0 +1,199 @@
+  return BindOnce([]() -> absl::optional<std::string> {
+    return "3669677 (name with )( ) R 477564 3669677 477564 34817 3669677 "
+           "4194304 91 0 0 0 1 3 0 0 20 0 1 0 163348716 9076736 384 "
+           "18446744073709551615 94595846197248 94595846217801 140722879734192 "
+           "0 0 0 0 0 0 0 0 0 17 52 0 0 0 0 0 94595846237232 94595846238880 "
+           "94595860643840 140722879736274 140722879736294 140722879736294 "
+           "140722879741931 0";
+  });
+}
+
+TEST_F(ProcessMetricsHelperTest, GetProcStatFields) {
+  auto fields = GetProcStatFields(GetStatCallback(), {0, 1, 2, 21, 51});
+  EXPECT_EQ(5, fields.size());
+  EXPECT_EQ("3669677", fields[0]);
+  EXPECT_EQ("name with )( ", fields[1]);
+  EXPECT_EQ("R", fields[2]);
+  EXPECT_EQ("163348716", fields[3]);
+  EXPECT_EQ("0", fields[4]);
+
+  fields = GetProcStatFields(GetStatCallback(), {0});
+  EXPECT_EQ(1, fields.size());
+  EXPECT_EQ("3669677", fields[0]);
+
+  fields = GetProcStatFields(GetStatCallback(), {0, 52});
+  EXPECT_EQ(0, fields.size());
+
+  fields = GetProcStatFields(GetNulloptCallback(), {0});
+  EXPECT_EQ(0, fields.size());
+}
+
+TEST_F(ProcessMetricsHelperTest, GetClockTicksPerSWithCallbacks) {
+  EXPECT_EQ(0, GetClockTicksPerS(GetNulloptCallback(), GetNulloptCallback()));
+  EXPECT_EQ(0, GetClockTicksPerS(GetNulloptCallback(), GetStatCallback()));
+  EXPECT_EQ(0, GetClockTicksPerS(GetUptimeCallback(), GetNulloptCallback()));
+  EXPECT_EQ(100, GetClockTicksPerS(GetUptimeCallback(), GetStatCallback()));
+}
+
+TEST_F(ProcessMetricsHelperTest, GetClockTicksPerS) {
+  EXPECT_EQ(100, ProcessMetricsHelper::GetClockTicksPerS());
+}
+
+TEST_F(ProcessMetricsHelperTest, GetCumulativeCPUUsage) {
+  TimeDelta usage = ProcessMetricsHelper::GetCumulativeCPUUsage();
+  EXPECT_GE(usage.InMicroseconds(), 0);
+}
+
+TEST_F(ProcessMetricsHelperTest, GetCPUUsage) {
+  TimeDelta usage = GetCPUUsage(GetStatCallback(), 100);
+  EXPECT_EQ(40, usage.InMilliseconds());
+}
+
+TEST_F(ProcessMetricsHelperTest, GetCumulativeCPUUsagePerThread) {
+  base::Value cpu_per_thread =
+      ProcessMetricsHelper::GetCumulativeCPUUsagePerThread();
+  base::Value::List* list = cpu_per_thread.GetIfList();
+  int initial_num_threads = list ? list->size() : 0;
+
+  Thread thread1("thread1");
+  Thread thread2("thread2");
+  Thread thread3("thread3");
+
+  thread1.StartAndWaitForTesting();
+  thread2.StartAndWaitForTesting();
+  thread3.StartAndWaitForTesting();
+
+  ASSERT_TRUE(thread1.IsRunning());
+  ASSERT_TRUE(thread2.IsRunning());
+  ASSERT_TRUE(thread3.IsRunning());
+
+  std::vector<std::string> vec1;
+  std::vector<std::string> vec2;
+  std::vector<std::string> vec3;
+
+  keep_busy.store(true);
+  thread1.task_runner()->PostTask(FROM_HERE, BindOnce(&BusyWork, &vec1));
+  thread2.task_runner()->PostTask(FROM_HERE, BindOnce(&BusyWork, &vec2));
+  thread3.task_runner()->PostTask(FROM_HERE, BindOnce(&BusyWork, &vec3));
+  PlatformThread::Sleep(Milliseconds(50));
```

Comment:
This may be flaky. There is a PostTaskAndReply, that calls the reply after the task has run, and you can use a WaitableEvent to wait for the reply to be called. Although you may have to use three of those, or a semaphore to be sure that all threads have started.